codec/encode.go

```diff
@@ -565,70 +604,101 @@ func (e *Encoder) kStruct(f *codecFnInfo, rv reflect.Value) {
 
 	var spool *sync.Pool
 	var poolv interface{}
-	var fkvs []stringRv
+	var fsfs []structField
 	// fmt.Printf(">>>>>>>>>>>>>> encode.kStruct: newlen: %d\n", newlen)
 	if newlen <= 8 {
-		spool, poolv = pool.stringRv8()
-		fkvs = poolv.(*[8]stringRv)[:newlen]
+		spool, poolv = pool.structField8()
+		fsfs = poolv.(*[8]structField)[:newlen]
 	} else if newlen <= 16 {
-		spool, poolv = pool.stringRv16()
-		fkvs = poolv.(*[16]stringRv)[:newlen]
+		spool, poolv = pool.structField16()
+		fsfs = poolv.(*[16]structField)[:newlen]
 	} else if newlen <= 32 {
-		spool, poolv = pool.stringRv32()
-		fkvs = poolv.(*[32]stringRv)[:newlen]
+		spool, poolv = pool.structField32()
+		fsfs = poolv.(*[32]structField)[:newlen]
 	} else if newlen <= 64 {
-		spool, poolv = pool.stringRv64()
-		fkvs = poolv.(*[64]stringRv)[:newlen]
+		spool, poolv = pool.structField64()
+		fsfs = poolv.(*[64]structField)[:newlen]
 	} else if newlen <= 128 {
-		spool, poolv = pool.stringRv128()
-		fkvs = poolv.(*[128]stringRv)[:newlen]
+		spool, poolv = pool.structField128()
+		fsfs = poolv.(*[128]structField)[:newlen]
 	} else {
-		fkvs = make([]stringRv, newlen)
+		fsfs = make([]structField, newlen)
 	}
 
 	newlen = 0
-	var kv stringRv
+	var sf structField
 	recur := e.h.RecursiveEmptyCheck
-	sfn := structFieldNode{v: rv, update: false}
+	sfn := structFieldNode{v: rvs, update: false}
 	for _, si := range tisfi {
-		// kv.r = si.field(rv, false)
-		kv.r = sfn.field(si)
+		sf.known = true
+		// sv.fal = si.field(rvs, false)
+		sf.val = sfn.field(si)
 		if toMap {
-			if si.omitEmpty() && isEmptyValue(kv.r, e.h.TypeInfos, recur, recur) {
+			if si.omitEmpty() && isEmptyValue(sf.val, e.h.TypeInfos, recur, recur) {
 				continue
 			}
-			kv.v = si.encName
+			sf.name = si.encName
 		} else {
 			// use the zero value.
 			// if a reference or struct, set to nil (so you do not output too much)
-			if si.omitEmpty() && isEmptyValue(kv.r, e.h.TypeInfos, recur, recur) {
-				switch kv.r.Kind() {
+			if si.omitEmpty() && isEmptyValue(sf.val, e.h.TypeInfos, recur, recur) {
+				switch sf.val.Kind() {
 				case reflect.Struct, reflect.Interface, reflect.Ptr, reflect.Array, reflect.Map, reflect.Slice:
-					kv.r = reflect.Value{} //encode as nil
+					sf.val = reflect.Value{} //encode as nil
 				}
 			}
 		}
-		fkvs[newlen] = kv
+		fsfs[newlen] = sf
+		newlen++
+	}
+
+	for name, encodedVal := range ufs.fields {
+		sf.name = name
+		sf.known = false
+		sf.encodedVal = encodedVal
+		fsfs[newlen] = sf
 		newlen++
 	}
 
 	if toMap {
+		if len(ufs.fields) > 0 {
+			// We have unknown fields, so we need to
+			// re-sort. Also check for duplicate field
+			// names.
+			sort.Sort(structFieldSlice(fsfs[:newlen]))
+			for j := 1; j < newlen; j++ {
```

Comment:
Seems like this loop is just for sanity-checking, and upstream might object to it over performance concerns.

Reply:
Perhaps...we're already iterating through the fields below, so maybe if they do object I can fold the check into that loop.